Add run_once task keyword to cause a task to bypass the host loop

sivel opened the pull request:
This pull request adds a `bypass` task key, to allow a task to funciton like a bypass host loop plugin similar to add_host/pause/etc...

This was inspired by:
https://groups.google.com/forum/#!topic/ansible-project/3-LaugNHe0M

Example:
###### bypass.yml

```

---
- hosts: all
  tasks:
    - debug:
        var: inventory_hostname
      bypass: True
```
###### output

```
PLAY [all] ********************************************************************

GATHERING FACTS ***************************************************************
ok: [localhost]
ok: [srv1.example.org]
ok: [srv2.exampe.org]
ok: [srv3.example.org]

TASK: [debug ] ****************************************************************
ok: [localhost] => {
    "inventory_hostname": "localhost"
}

PLAY RECAP ********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
srv1.example.org           : ok=2    changed=0    unreachable=0    failed=0
srv2.example.org           : ok=2    changed=0    unreachable=0    failed=0
srv3.example.org           : ok=2    changed=0    unreachable=0    failed=0
```

sivel committed:
Add run_once task key

This caused the task to act like  bypass host loop plugin and
execute only on a single host. Can be used with delegate_to

mpdehaan commented:
I can definitely see value for this though most of my question is in how we present it.

Should this maybe be something like "just_once: True" or even "delegate_once: localhost" or even "run_only: N" (times) (I thik I like this last one less)

I'm not entirely sure, though in either case we'll need to add this to an appropriate docs section -- it's a bit of an advanced concept and I don't want to clutter the main paths with the complex workarounds, but again, I can see this being useful.

Thoughts?

sivel commented:
I did look for an existing docs section and did not find one.  So I think we might need to add one.

I'm open to modification.  I was just sticking with existing (internal) terms already in use, for a true/false configuration.

But looking at your suggestion I do like "delegate_once" and specifying the host to delegate to.

sivel commented:
I am wondering, if we could still leave it as 'bypass=true/false' and pair it with 'delegate_to' to achieve 'delegate_once' as opposed to a actually creating a delegate_once option.

But I do think that 'bypass' might be misleading, so something along the lines of 'just_once' sounds good.  I just am not sure I like the name 'just_once'.

mpeters commented:
I'm definitely in favor of this feature, but I'm not sold on the "bypass" keyword. That makes it seems like it's going to be skipped completely. I also don't like "delegate_once" because then you have to specify the server to run on it. If we had a "just_once" flag then it could be combined with a "delegate" option to achieve that. So I like the semantics that just_once implies even though that name seems a little weak.

BTW, as far as where in the documentation this would live, I'd suggest the docs on rolling upgrades/deployment as that's where serial is discussed as well and they seem similar in "advancedness". Plus I think this feature is useful for deployment.

mpeters commented:
Some other options for the name:
- first_host
- single_host
- run_once
- execute_once
- host_limit (more like run_only)

jimi-c commented:
My $0.02 is for `single_host: <hostname>`, with optional syntax for specifying the first or last host in the group.

mpeters commented:
I was just thinking `single_host: true` not with an actual hostname. If you wanted that you could do `single_host: true` and `delegate_to: <hostname>`. I don't think `single_host` should do what `delegate_to` does as well.

sivel commented:
My only "concern" with `single_host: <hostname>` is needing to then handle also specifying "delegate_to", or even "local_action".

Just some things that need to be taken into account.

I think if we leverage existing local_action and delegate_to functionality paired with some type of "only run me once" option, that it reduces points of confusion.

```
- debug: var=inventory_hostname
  single_host: true
  delegate_to: foobar
```

If delegate_to is not present, then it functions like add_host, and just does hosts[0].

However, I think the wording of "single_host" then seems strange when using "delegate_to".

mpeters commented:
But I could see allowing `single_host: true`, and `single_host: first` or `single_host: last`.

mpeters commented:
I agree with @sivel. Having composable features is good. So the parameter should probably not have `host` in the name. So something like `run_once`, etc.

jimi-c commented:
They could be made mutually exclusive, as su and sudo are now, but I'm fine with making them composable as you say. In that case it should be just `single_host: yes` (or whatever boolean string you prefer) and let delegate_to/local_action define the target host - the default being to act on the first host in the list as BYPASS does now.

bcoca commented:
do we really need extra semantics for what can already be done ?

when: inventory_hostname == play_hosts[0]

it is not as pretty,  but I find that adding semantics that don't simplify
things by a lot just complicates the tool needlessly.

mpeters commented:
I can see your point @bcoca if that's all it did. But I was hoping that it would also be an option available to whole plays so you didn't have to litter multiple tasks with it.

Also, if you have dozens or hundreds of machines in a group, then that's a lot of skips to wait on and fill the screen with. If the pause module can do it, it would be nice it other tasks could too without having to write a whole custom module for it.

But I also understand that littering a big generic project like ansible with lots of custom nice-to-haves is problematic. I'm just hoping this is useful enough to overcome that threshold.

jimi-c commented:
@bcoca we had discussed this, where you could achieve this using "with" and "delegate_to", however that's a lot of boilerplate that might need to be repeated endlessly, so we thought it'd be worth-while to have a language feature for it specifically.

sivel commented:
I'll start working on the improvements for this soon and post updates as I progress.

sivel commented:
I've renamed `bypass` to `run_once` and squashed and pushed.

This functionality will already allow use with `delegate_to` without any additional modifications.

```
---
- hosts: all
  gather_facts: False
  tasks:
    - command: hostname
      run_once: True

    - command: hostname
      run_once: True
      delegate_to: srv3.example.org
```

```
PLAY [all] ********************************************************************

TASK: [command hostname] ******************************************************
changed: [localhost] => {"changed": true, "cmd": ["hostname"], "delta": "0:00:00.003770", "end": "2014-05-19 11:10:27.091348", "rc": 0, "start": "2014-05-19 11:10:27.087578", "stderr": "", "stdout": "localhost.localdomain"}

TASK: [command hostname] ******************************************************
changed: [localhost] => {"changed": true, "cmd": ["hostname"], "delta": "0:00:00.008851", "end": "2014-05-19 11:10:27.446403", "rc": 0, "start": "2014-05-19 11:10:27.437552", "stderr": "", "stdout": "srv3.example.org"}

PLAY RECAP ********************************************************************
localhost                  : ok=2    changed=2    unreachable=0    failed=0
srv1.example.org           : ok=2    changed=2    unreachable=0    failed=0
srv2.example.org           : ok=2    changed=2    unreachable=0    failed=0
srv3.example.org           : ok=2    changed=2    unreachable=0    failed=0
```

The delegation portion of the playbooks documentation has been updated to include a 'run_once' section.

This isn't supported at the play level (currently).  I'm not sure it's needed there, but if someone thinks it should be, I can always look into it.  I figure the "hosts" specification should just be specified appropriately to only target the host you want to execute on.

iamjvn commented:
I have a scenario where this will be useful in a rolling upgrade playbook. 
- I currently use "when: inventory_hostname == play_hosts[0]" on every task in a "pre play" to run those cluster pre-configuration tasks only once from only one of the servers in the group. 
- This "pre play" is then followed by the actual rolling upgrade play touching every server in the group.
- Finally, I use "when: inventory_hostname == play_hosts[0]" again, on every task in a "post play" to run those cluster post-configuration tasks only once from only one of the servers in the group. 

Unless there's a better way..

hoettle commented:
+1 re the use of pre / post play tasks..especially for singleton actions such as DB schema upgrades or some other non-repeated activity. A run_once keyword would definitely improve readability in this case.

danillos commented:
+1

jimi-c commented:
Merged, thanks!

jimsmith commented:
Hello there I have this running ansible 1.7 and run_once does not work.

I have a DB schema update and other one time only tasks that need to run, I am also using delegate_to and serial as well so it and the playbook honors the delegate_to and serial set as 1 but the run_once tasks are still being ran more than 'once' on the delegate_to host that I've assigned.

```
 - name: run django migrations
  command: "/path/to/db_migrations.sh"
  # run_once: true -- this action will be availabe in ansible version 1.7, which is dev version now.
  # run_once doesn't work in current release, so 'delegate_to' should be specified to be executed for particular host
  run_once: True
  delegate_to: "192.168.1.2"      
  register: run_migrations
  # Print all contents of the shell task's output.
  - debug: var=run_migrations.stdout_lines
```

So the above is not being honored fully, as I have 2 hosts and the command gets ran more than once.

?
